library/std/src/sys/fs/windows.rs

```diff
@@ -1233,74 +1233,147 @@ pub fn unlink(path: &WCStr) -> io::Result<()> {
 }
 
 pub fn rename(old: &WCStr, new: &WCStr) -> io::Result<()> {
-    if unsafe { c::MoveFileExW(old.as_ptr(), new.as_ptr(), c::MOVEFILE_REPLACE_EXISTING) } == 0 {
-        let err = api::get_last_error();
-        // if `MoveFileExW` fails with ERROR_ACCESS_DENIED then try to move
-        // the file while ignoring the readonly attribute.
-        // This is accomplished by calling `SetFileInformationByHandle` with `FileRenameInfoEx`.
-        if err == WinError::ACCESS_DENIED {
-            let mut opts = OpenOptions::new();
-            opts.access_mode(c::DELETE);
-            opts.custom_flags(c::FILE_FLAG_OPEN_REPARSE_POINT | c::FILE_FLAG_BACKUP_SEMANTICS);
-            let Ok(f) = File::open_native(&old, &opts) else { return Err(err).io_result() };
-
-            // Calculate the layout of the `FILE_RENAME_INFO` we pass to `SetFileInformation`
-            // This is a dynamically sized struct so we need to get the position of the last field to calculate the actual size.
-            let Ok(new_len_without_nul_in_bytes): Result<u32, _> =
-                ((new.count_bytes() - 1) * 2).try_into()
-            else {
-                return Err(err).io_result();
+    // Calculate the layout of the `FILE_RENAME_INFO` we pass to `SetFileInformation`
+    // This is a dynamically sized struct so we need to get the position of the last field to calculate the actual size.
+    let Ok(new_len_without_nul_in_bytes): Result<u32, _> = ((new.count_bytes() - 1) * 2).try_into()
+    else {
+        return Err(WinError::INVALID_PARAMETER).io_result();
```

Comment:
Probably `ERROR_FILENAME_EXCED_RANGE` would be better or else a custom error message. Even if it's not strictly what happened before.